Add blog post #3: Quarkus Unpacked - Foojay Podcast Companion (#2788)
quarkusio/quarkusio.github.io: _posts/2026-06-23-mmaler-blogpost-3-quarkus-unpacked-foojay-podcast.adoc

synopsis: A companion to the Foojay podcast episode, this post distills the key questions and answers about Quarkus — from build-time optimization and live reload to JVM versus native trade-offs and cloud cost savings.
author: mmaler
thumbnailimage: /assets/images/posts/mmaler-blog-posts/quarkus-unpacked-foojay-podcast.jpeg
---
:imagesdir: /assets/images/posts/mmaler-blog-posts/3-quarkus-unpacked-foojay-podcast/
ifdef::env-github,env-browser,env-vscode[:imagesdir: ../assets/images/posts/mmaler-blog-posts/3-quarkus-unpacked-foojay-podcast/]

= Quarkus Unpacked: Insights from the Foojay Podcast

image::/assets/images/posts/mmaler-blog-posts/quarkus-unpacked-foojay-podcast.jpeg[alt="Quarkus Unpacked: Insights from the Foojay Podcast", width=100%]

I recently had the pleasure of joining the link:https://www.youtube.com/watch?v=_nJCTTrnZkE[Foojay podcast] to talk about Quarkus in depth.
The conversation covered a lot of ground, from what makes Quarkus different to the practical trade-offs between JVM and native mode.
This post captures the key questions and answers from that discussion, lightly edited for readability.

> If you have been following this blog series, note that the third installment on building your own stack with Quarkus is coming next.

Reviewer: Should that be fourth installment? If this is the third?

Author: Nice one. Good catch.

Consider this a bonus entry that distills the podcast conversation into a format you can read, reference, and share.

== What is Quarkus?

Quarkus is a cloud-native Java framework for any cloud, hyperscalers included, that combines efficiency and cost savings with a great developer experience.
It features innovations such as Dev Mode and Continuous Testing, while it keeps the enterprise stability that Java is known for.

Quarkus is designed for cloud-native Java developers who need quick cold starts, high-density deployments, and an efficient workflow with live reload.
At the same time, it keeps standard APIs and offers the option of running on the JVM or in native mode.

== How does Quarkus compare to Spring, Micronaut, or other frameworks?

Broadly speaking, Quarkus, Spring Boot, and Micronaut all serve the enterprise market, but Quarkus stands out for its focus on cloud-native defaults.

> While Spring Boot has traditionally relied on runtime reflection and classpath scanning, it has been adding AOT processing support in recent versions. Quarkus, by contrast, was designed from the start around build-time processing and native compilation with GraalVM, which results in faster startup times and reduced memory usage.

Reviewer: I wonder if we should say "enabling native compilation" to make it clear you can run it in jvm mode?

Author: Yep, applying. Thank you, @holly-cummins !

From a day-to-day perspective, Quarkus also enhances the developer experience with features like Dev Mode and live reload, providing instant feedback on code changes with no long rebuilds or restarts.

On top of that, Quarkus integrates both imperative and reactive models by using Eclipse Vert.x, and it still adheres to solid standards like MicroProfile and Jakarta EE.

== Is Quarkus more modern because it is newer?

Quarkus is newer than many established Java frameworks, but "newer" does not automatically mean "more modern."
In practice, its design avoids legacy constraints and stays away from older patterns that do not fit cloud-native deployments.
Instead, Quarkus focuses on build-time processing and a streamlined developer loop.

It also applies lessons from earlier frameworks, and you can see that in its efficient extension model and configuration.
Quarkus feels more modern not just because it is newer, but because it embraces cloud paradigms at a fundamental level.

== Does Quarkus replace the JVM?

No, it does not replace the JVM.

> In JVM mode, Quarkus runs on the JVM like other Java frameworks, while leveraging its GraalVM integration.

Reviewer: ```suggestion
In JVM mode, Quarkus runs on the JVM like other Java frameworks, while leveraging its build-time optimizations.
```

Instead of doing this work every time the application starts, Quarkus does it once during compilation and generates optimized bytecode that runs with precomputed results.

This gives you four main benefits:

* **Faster startup**, because the application skips the heavy lifting of scanning and parsing when it boots.
* **Lower memory usage**, because the final artifact stays smaller and leaner, and libraries needed only for initialization, such as parsers, do not get loaded at runtime.
* **Earlier error detection**, because Quarkus catches configuration errors or incorrect annotation usage during the build, so it fails fast instead of crashing at runtime.
* **Reduced reflection**, because Quarkus generates proxies that avoid reflection calls, which improves performance.

=== How does this differ from JIT and AOT?

It helps to separate build-time optimization from both just-in-time (JIT) compilation and ahead-of-time (AOT) initiatives like Project Leyden.

With JIT, the JVM optimizes hot code paths into machine code while the application runs.

> Quarkus build-time optimizations do something different: they improve the baseline startup and memory profile, so the application reaches the JIT warm-up phase faster.

Reviewer: ```suggestion
Quarkus build-time optimizations do something different: they improve the baseline startup and memory profile, so the application reaches the warmed JIT phase faster.
```

* The **Development UI** (Dev UI) runs alongside the app and displays configuration, active beans, and routing, so troubleshooting is more direct.
* For **Kubernetes-native deployment**, Quarkus can generate Kubernetes manifests and build container images by using Jib or Docker, so developers can deploy without learning every Kubernetes detail first.

=== Day 2: Operations

* For **security**, Quarkus integrates standards such as OIDC (OpenID Connect) and WebAuthn (passwordless authentication) through dedicated extensions. Because Quarkus does framework work at build time, it can reduce the deployed attack surface by removing unused code paths before deployment.
* For **observability** (metrics, logs, and traces), Quarkus integrates with OpenTelemetry and Micrometer, so metrics, distributed tracing, and log correlation often work with defaults and require no extra setup.
* For **fault tolerance**, MicroProfile Fault Tolerance lets developers add annotations that apply circuit breakers, retries, or bulkheads, so the application handles common network failures without custom plumbing.

=== The Quarkiverse ecosystem

> Quarkus has the Quarkiverse ecosystem, a community extension catalog that provides a large set of extensions connecting Quarkus to common Java libraries and platforms.

Reviewer: ```suggestion
Quarkus has the Quarkiverse ecosystem, a community extension catalog that provides a large set of extensions integrating Quarkus with common Java libraries and platforms.
```
